// osrs-hiscores/lib/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BOUNTY_HUNTER = ['hunter', 'rogue'];
exports.CLUES = ['all', 'beginner', 'easy', 'medium', 'hard', 'elite', 'master'];
exports.BOSSES = [
    'abyssalSire',
    'alchemicalHydra',
    'barrowsChests',
    'bryophyta',
    'callisto',
    'cerberus',
    'chambersOfXeric',
    'chambersOfXericChallengeMode',
    'chaosElemental',
    'chaosFanatic',
    'commanderZilyana',
    'corporealBeast',
    'crazyArchaeologist',
    'dagannothPrime',
    'dagannothRex',
    'dagannothSupreme',
    'derangedArchaeologist',
    'generalGraardor',
    'giantMole',
    'grotesqueGuardians',
    'hespori',
    'kalphiteQueen',
    'kingBlackDragon',
    'kraken',
    'kreeArra',
    'krilTsutsaroth',
    'mimic',
    'nightmare',
    'obor',
    'sarachnis',
    'scorpia',
    'skotizo',
    'theGauntlet',
    'theCorruptedGauntlet',
    'theatreOfBlood',
    'thermonuclearSmokeDevil',
    'tzKalZuk',
    'tzTokJad',
    'venenatis',
    'vetion',
    'vorkath',
    'wintertodt',
    'zalcano',
    'zulrah',
];
exports.SKILLS = [
    'overall',
    'attack',
    'defence',
    'strength',
    'hitpoints',
    'ranged',
    'prayer',
    'magic',
    'cooking',
    'woodcutting',
    'fletching',
    'fishing',
    'firemaking',
    'crafting',
    'smithing',
    'mining',
    'herblore',
    'agility',
    'thieving',
    'slayer',
    'farming',
    'runecraft',
    'hunter',
    'construction',
];
exports.ACTIVITIES = ['leaguePoints', ...exports.BOUNTY_HUNTER, ...exports.CLUES, 'lastManStanding', ...exports.BOSSES];
exports.MODES = ['normal', 'ironman', 'ultimate', 'hardcore', 'deadman', 'seasonal', 'tournament'];
